### DISCUSSION TITLE
Alternative source for US$ values than using current USDC price on Algorand

### DISCUSSION THREAD

**Discussion by ipaleka** [asastats/channel#53] [General]
This relates to the following feature request:

https://discord.com/channels/906917846754418770/908054304332603402/971201775904370778
> Low Priority Request: Wrapped asset fiat values would be better presented as their off chain equivalent USD value rather than the on chain liquidation value. AlgoFi has SDKs available to query their Oracle that could be leveraged to accomplish this

So basically this discussion is about two topics:

a) an alternative source for fetching ALGO price in US$
 
  It is worth mentioning that we'd need the same or another source for the internalization purposes (adding other world's major and not so major fiat currencies), but we can probably just use US$ value to calculate those.

b) wrapped coins' price information from the other blockchains rather than from Algorand AMMs

 Or to say from the other reliable offchain sources instead of using the same process used for all the other tokens on Algorand. Actually this point demands discussion as implementing this implies that clicking ALGO-USD switch button no more represents "translation", it's rather a complete change in approach for wrapped coins.

**Comment by wxivd** [comment]
Being pragmatic about how price information is used; Purchasing, selling, trading are all being fulfilled on Algorand blockchain. Thus price information from Algorand AMMs is ideal and facilitates the objectives.

Bridges to other blockchains are yet to be developed. Currently any wrapped asset can be exchanged on Algorand Dexs only.

Price information from the other blockchains would be essential once moving assets between the other blockchains is functional.

**Reply by ipaleka** [reply]
This is similar to my stance on this regarding point a).

USDC is currently the best approach and also we have an option to switch to some other stable coin, like Tether if there would be problems with USDC, but we kind of should take a stance that Algorand blockchain is always a basis.

Regarding point b), the requester sort of brings valid points there. In the previous case, if USDc or Tether can't cope with the problems then the shit hit the fan and ASA Stats can't do anything about, and using other sources besides onchain values won't fix anything, and it would be somehow misleading for our users.

But in a case of wrapped Bitcoin and other wrapped coins, there could be issues with liquidity - and that's exactly what caused this feature request - and the right values aren't shown to the users. The fact is that they can exchange their wrapped coins 1:1 and then our calculations based on liquidity are proven to be wrong.

I mean, the most frequent question in ASA Stats has been that FAQ entry "Why my value isn't my amount multiplied with the price on Tinychart?" (when they hold plenty of the related tokens) and our valid and right answer is "Because you can't get that much.", but in this wrapped coin case we can't answer the same.

**Reply by wxivd** [reply]
> "Because you can't get that much.", but in this wrapped coin case we can't answer the same.

What about giving a price valuation based **Not** on how much it can be exchanged for on AMMs, but on how much (Algo) lending platforms (i.e Algofi) would give incase the user opt-ed to lend wrapped asset. (Mathematically account for borrow utilization, that is cancelling out the max borrow to get real value). 

> Borrow utilization on [Algofi](https://app.algofi.org/lend) "Your max borrow is determined by your supplied assets. The increase in maximum borrow versus amount deposited varies by asset. You are limited to borrowing 90% of your max borrow"

In theory, if this works, then we can provide two values. The current value based on AMMs and the new value which is expected to be higher as it is not prone to low-liquidity issues.

**Reply by ipaleka** [reply]
> > "Because you can't get that much.", but in this wrapped coin case we can't answer the same.
> 
> What about giving a price valuation based **Not** on how much it can be exchanged for on AMMs, but on how much (Algo) lending platforms (i.e Algofi) would give incase the user opt-ed to lend wrapped asset. (Mathematically account for borrow utilization, that is cancelling out the max borrow to get real value).

This would create two problems out of one. Our problem is only in US$ and only for wrapped coins. We can either do nothing, or define wrapped coins and their US$ price sources so their value in US$ for the case where they exchange wrapped coins in 1:1 ratio for coins and then sell those coins on a major crypto exchange.

P.S. Afaik people are able to swap their goBTC and goETH to BTC and ETH on https://www.algomint.io/.

**Comment by ipaleka** [comment]
by [u/jasonl999](https://www.reddit.com/user/jasonl999/): [Borrowed goETH value is wrong](https://www.reddit.com/r/asastats/comments/ut1dl6/borrowed_goeth_value_is_wrong/)

- - - - -

First off, I love ASAStats, this is not a complaint just feedback!

This problem is magnified quite a bit by the fact that the swap rate is significantly better at [pact.fi](https://pact.fi/), but here's the issue:

If I've borrowed 1 goETH, the "true" value is the cost of swapping from algo into goETH to pay it back. Right now, there's about a 40% price impact on tinyman; 1 goETH ($1,948) gets you 3,208 algo ($1,400). asastats treats that as the negative value of algo.

The problem is it should be doing the swap in the other direction when determining portfolio value in terms of algo: "how many ago do I need to get 1 goETH so I can pay it back". In this case it would take 7,312 algo ($3,192) to pay back 1 goETH.

This problem is not as bad if [pact.fi](https://pact.fi/) is used to swap since they have much higher liquidity (it would take 4,486 algo worth $1,953), but even in that case, the evaluation for portfolio purposes should be "how much algo would it take to pay back this loan".

Hope that explains it clearly!

- - - - -

**Reply by ipaleka** [reply]
We had a similar discussion on our Discord last year and it somehow went without an outcome, iirc a similar argument as this was brought for borrowed funds. Anyway, we should at least have a discussion with an outcome this time, in the worst-case scenario at least an entry on our FAQ page...

**Reply by ipaleka** [reply]
by u/jasonl999 on [Reddit](https://www.reddit.com/r/asastats/comments/uwvhsj/comment/i9xiacg/?utm_source=share&utm_medium=web2x&context=3):
> Looks good! Can confirm my issue where I have borrowed goETH that's in a liquidity pool now calculates the net value correctly, awesome work.
